pyro/mesh/patch.py

```diff
@@ -885,5 +885,174 @@ def do_demo():
+        ymax : float, optional
+            Physical coordinate at the upper y boundary
+        """
+
+        # pylint: disable=too-many-arguments
+
+        # size of grid
+        self.nx = int(nx)
+        self.ny = int(ny)
+        self.ng = int(ng)
+
+        self.qx = int(2*ng + nx)
+        self.qy = int(2*ng + ny)
+
+        # domain extrema
+        self.xmin = xmin
+        self.xmax = xmax
+
+        self.ymin = ymin
+        self.ymax = ymax
+
+        # compute the indices of the block interior (excluding guardcells)
+        self.ilo = self.ng
+        self.ihi = self.ng + self.nx-1
+
+        self.jlo = self.ng
+        self.jhi = self.ng + self.ny-1
+
+        # center of the grid (for convenience)
+        self.ic = self.ilo + self.nx//2 - 1
+        self.jc = self.jlo + self.ny//2 - 1
+
+        # define the coordinate information at the left, center, and right
+        # zone coordinates
+        self.dx = (xmax - xmin)/nx
+
+        self.xl = (np.arange(self.qx) - ng)*self.dx + xmin
+        self.xr = (np.arange(self.qx) + 1.0 - ng)*self.dx + xmin
+        self.x = 0.5*(self.xl + self.xr)
+
+        self.dy = (ymax - ymin)/ny
+
+        self.yl = (np.arange(self.qy) - ng)*self.dy + ymin
+        self.yr = (np.arange(self.qy) + 1.0 - ng)*self.dy + ymin
+        self.y = 0.5*(self.yl + self.yr)
+
+        # 2-d versions of the zone coordinates (replace with meshgrid?)
+        x2d = np.repeat(self.x, self.qy)
+        x2d.shape = (self.qx, self.qy)
+        self.x2d = x2d
+
+        y2d = np.repeat(self.y, self.qx)
+        y2d.shape = (self.qy, self.qx)
+        y2d = np.transpose(y2d)
+        self.y2d = y2d
+
+
+    def face_area(self):
+        """
+        Return an array of the face areas. 
+        The shape of the returned array is (ni, nj).
+        """
+        tr = lambda arr: arr.transpose(1, 2, 0)
+        x = self.cell_vertices()[:,0]
+        y = self.cell_vertices()[0,:]
+        r0 = x[:-1, :-1]
+        r1 = x[+1:, :-1]
+        t0 = y[:-1, :-1]
+        t1 = y[+1:, +1:]
+
+        # ** the area of the arc
+
+        area_i = np.pi * (r0 * np.sin(t0) + r0 * np.sin(t1)) * np.sqrt(np.square(r0 * np.sin(t1) - r0 * np.sin(t0)) + np.square(r0 * np.cos(t1) - r0 * np.cos(t0)))
+        area_j = np.pi * (r0 * np.sin(t0) + r1 * np.sin(t0)) * np.sqrt(np.square(r1 * np.sin(t0) - r0 * np.sin(t0)) + np.square(r1 * np.cos(t0) - r0 * np.cos(t0)))
+        return tr(np.array([area_i, area_j]))
+
+    def cell_volumes(self):
+        """
+        Return an array of the cell volume data for the given coordinate box
+        The shape of the returned array is (ni, nj).
+        """
+
+        x = self.cell_vertices()[:,0]
+        y = self.cell_vertices()[0,:]
+
+        r0 = x[:-1, :-1]
+        r1 = x[+1:, :-1]
+        t0 = y[:-1, :-1]
+        t1 = y[+1:, +1:]
+
+        return (r1 ** 3 - r0 ** 3) * (np.cos(t1) - np.cos(t0)) * (-2.0 * np.pi) / 3.0
+        # return r1
+
+    def cell_vertices(self):
+        """
```

Comment:
isn't this already what self.xl and self.xr provide?
:)

Reply:
I did not really understand what they are, and decided to write my own function. ;)